### PR TITLE
Uat list implementation

### PR DESCRIPTION
**Pull request by joshcbee**
List model and add/edit interface added, accessible through:
staff/create/list/ and staff/edit/list/?id=[num]

Addresses #219, #220 

You will need to migrate database for this.

Expected behaviour:
- Create a list of courses from those available in your database and save that list
- Reload that list, edit and save again

There are a couple of TODOs still in the code, wanting to avoid a massive PR and this seemed like a good piece to start with, will continue implementing other elements (e.g. program/subplan tie in, their own view page).

image

### CODE REVIEW

**Review by JackSekoranja — CHANGES_REQUESTED** [review]
Interface is pretty good, but I found that I wasn't able to save the list, even after importing dependencies and migrating (See screenshot). I also found courses appeared multiple times and were out of order, so if possible that would be good to fix. My search text also kept deleting itself whenever I added a list item, which kind of got annoying as I can see myself adding multiple courses in the same subject area (Not sure if this is intentional though). As a styling change, it would also be good to make the list appear like it is part of the text box above and not just floating below it.

Screen Shot 2019-08-28 at 1 35 22 pm

File: cassdegrees/api/models.py [hunk 1]
```diff
@@ -23,6 +23,18 @@ class Meta:
         unique_together = (("code", "year"),)
 
 
+# todo: Add type for elements
```
Comment (j-selby) [inline cassdegrees/api/models.py:26]: Make sure you open an issue for this one afterwards

File: cassdegrees/static/css/style.css [hunk 2]
```diff
@@ -378,3 +378,57 @@ select.text{
         display: block;
     }
 }
+
+/*todo remove*/
```
Comment (j-selby) [inline cassdegrees/static/css/style.css]: Ditto

File: cassdegrees/templates/createlist.html [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    {% else %}
+        {% finalcrumb 'Create List' %}
+    {% endif %}
+{% endblock %}
+
+{% block content %}
+
+    <script src="https://cdn.jsdelivr.net/npm/vue/dist/vue.js"></script>
+    <script src="https://unpkg.com/vue-multiselect@2.1.0"></script>
+    <script src="https://cdn.jsdelivr.net/npm/vue-resource@1.5.1"></script>
```
Comment (j-selby) [inline cassdegrees/templates/createlist.html]: scripts typically go at the end. These should also be bundled with the application to minimise cross domain requests - e.g. we are already shipping vue.

File: cassdegrees/templates/createlist.html [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+               onclick="submit_form(this.value, false)">
+        {% endif %}
+        <input class="btn-uni-grad btn-large" type="submit" value="{% if edit %}Save and Exit{% else %}Create{% endif %}"
+               onclick="submit_form(this.value, true)">
+    </p>
+
+    <script>
+        function submit_form(form_action, redirect) {
+            if (true) {
+                document.getElementById('mainForm').action.value = form_action;
```
Comment (j-selby) [inline cassdegrees/templates/createlist.html]: `setAttribute` should be used here instead.

File: cassdegrees/templates/createlist.html [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                document.getElementById('redirect').value = redirect;
+                document.getElementById('mainForm').submit();
+                return true;
+            } else {
+                return false;
+            }
+        }
+    </script>
+
+
```
Comment (j-selby) [inline cassdegrees/templates/createlist.html]: Un-needed spacing

File: cassdegrees/templates/list.html [hunk 6]
```diff
@@ -21,7 +21,7 @@
         <p class="msg-error">{{ render.error|linebreaksbr }}</p>
     {% endif %}
 
-    <form class="box bdr-solid bdr-uni bg-uni50 anuform custom_search_bar" action="?action=Search" method="get">
+    <form class="box bdr-solid bdr-uni bg-uni50 anuform custom_search_bar" action="/list/?action=Search" method="get">
```
Comment (j-selby) [inline cassdegrees/templates/list.html]: Should this be relative?

File: cassdegrees/templates/widgets/courselistingwidget.html [hunk 7]
```diff
@@ -0,0 +1,174 @@
+{% load static %}
+
+{# Usage: Must pass in a string app_name that is id of the particular widget the app is to be bound to#}
+{# Must include imports below once in the relevant template prior to insertion of a widget #}
+{#    <script src="https://cdn.jsdelivr.net/npm/vue/dist/vue.js"></script>#}
+{#    <script src="https://unpkg.com/vue-multiselect@2.1.0"></script>#}
+{#    <script src="https://cdn.jsdelivr.net/npm/vue-resource@1.5.1"></script>#}
+{#    <link rel="stylesheet" href="https://unpkg.com/vue-multiselect@2.1.0/dist/vue-multiselect.min.css">#}
```
Comment (j-selby) [inline cassdegrees/templates/widgets/courselistingwidget.html]: See issue above re: scripts

File: cassdegrees/templates/widgets/courselistingwidget.html [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+                    <i class="fa fa-times fa-fw"></i>
+                </div>
+            </li>
+        </template>
+    </ul>
+
+</div>
+
+<script type="application/javascript">
+
```
Comment (j-selby) [inline cassdegrees/templates/widgets/courselistingwidget.html:56]: This is likely too long to include inline - can this be moved to a .js file?

File: cassdegrees/ui/urls.py [hunk 9]
```diff
@@ -52,4 +53,6 @@
     path(staff_url_prefix + 'view/subplan/', view_),
     path(staff_url_prefix + 'view/course/', view_),
     path(staff_url_prefix + 'pdf/program/', view_program_pdf),
+    path(staff_url_prefix + 'create/list/', create_list),
```
Comment (j-selby) [inline cassdegrees/ui/urls.py]: Can you sort this into the other categories?

File: cassdegrees/ui/views/lists.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        else:
+            form = EditListFormSnippet()
+
+    return render(request, 'createlist.html', context={
+        "edit": False,
+        "form": form,
+    })
+
+
+# Todo: implement list deletion with associated business rules (if any)
```
Comment (j-selby) [inline cassdegrees/ui/views/lists.py:57]: Issue for this one as well if possible after merge

**Review by j-selby — CHANGES_REQUESTED** [review]
Looks good, with a few issues.

This doesn't appear to run on Linux right now (with Python 3.7.4):

```
/home/james/.local/lib/python3.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)
Traceback (most recent call last):
  File "manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/home/james/.local/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/home/james/.local/lib/python3.7/site-packages/django/core/management/__init__.py", line 357, in execute
    django.setup()
  File "/home/james/.local/lib/python3.7/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/home/james/.local/lib/python3.7/site-packages/django/apps/registry.py", line 89, in populate
    app_config = AppConfig.create(entry)
  File "/home/james/.local/lib/python3.7/site-packages/django/apps/config.py", line 90, in create
    module = import_module(entry)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'corsheaders'
```

File: cassdegrees/ui/urls.py (hunk 9 above)
Comment (joshcbee) [reply cassdegrees/ui/urls.py]: Reordered listing.

File: cassdegrees/templates/widgets/courselistingwidget.html (hunk 8 above)
Comment (joshcbee) [reply cassdegrees/templates/widgets/courselistingwidget.html:56]: The benefit, (I think), of including it in a widget template is that the template calling the widget can dynamically pass it new app names (e.g. if multiple entry fields are required on the same page), and the widget template then creates a new vue object for each. I haven't found a way to be able to replicate this separating the file into js. However, all that needs to be included in a calling template is {% include ...widget... app_name %}, so this code is only located in the widget file. Happy to discuss further if you have any ideas?

File: cassdegrees/templates/createlist.html (hunk 3 above)
Comment (joshcbee) [reply cassdegrees/templates/createlist.html]: Have bundled the scripts, however the scripts must be declared before the first widget listing or the console will throw a "vue not defined" error.

File: cassdegrees/api/models.py (hunk 1 above)
Comment (joshcbee) [reply cassdegrees/api/models.py:26]: Following discussion at 28/8 team meeting, list structure not to be generalised at this point. Todo will be removed.

File: cassdegrees/templates/list.html (hunk 6 above)
Comment (joshcbee) [reply cassdegrees/templates/list.html]: I'm not sure where this has come from, I don't have in on my local as a change that I have made?

File: cassdegrees/ui/views/lists.py (hunk 10 above)
Comment (joshcbee) [reply cassdegrees/ui/views/lists.py:57]: Issue created

File: cassdegrees/templates/widgets/courselistingwidget.html (hunk 8 above)
Comment (j-selby) [reply cassdegrees/templates/widgets/courselistingwidget.html:56]: Ah, yep, didn't see that. Thats totally fine then.

**Review by JackSekoranja — CHANGES_REQUESTED** [review]
Still having issues, see screenshot below:

Screen Shot 2019-09-02 at 2 20 34 pm

**Comment by joshcbee** [comment]
@JackSekoranja could you check you have the latest pull and have cleared caches and try again, this should have been resolved

File: cassdegrees/static/css/style.css (hunk 2 above)
Comment (joshcbee) [reply cassdegrees/static/css/style.css]: reference added and todo removed

File: cassdegrees/templates/list.html (hunk 6 above)
Comment (joshcbee) [reply cassdegrees/templates/list.html]: Matched to develop

File: cassdegrees/ui/views/lists.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+from django.http import HttpResponseNotFound, HttpRequest
+from django.shortcuts import render, redirect
+from django.utils import timezone
+
+from ui.forms import EditListFormSnippet
+
+admin_url_prefix = "/staff/"
+list_program_url = admin_url_prefix + "list/?view=Program"
+
+list_course_url = admin_url_prefix + "list/?view=Course"
```
Comment (JackSekoranja) [inline cassdegrees/ui/views/lists.py]:
```suggestion
```

File: cassdegrees/ui/views/lists.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+from api.views import search
+
+from django.http import HttpResponseNotFound, HttpRequest
+from django.shortcuts import render, redirect
+from django.utils import timezone
+
+from ui.forms import EditListFormSnippet
+
+admin_url_prefix = "/staff/"
+list_program_url = admin_url_prefix + "list/?view=Program"
```
Comment (JackSekoranja) [inline cassdegrees/ui/views/lists.py]:
```suggestion
list_course_group_url = admin_url_prefix + "list/?view=List"
```

File: cassdegrees/ui/views/lists.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        # Find the list to specifically create from:
+        instance = ListModel.objects.get(id=int(id))
+
+    if request.method == 'POST':
+        form = EditListFormSnippet(request.POST)
+
+        # todo: implement list view on admin view page then change url
+        if form.is_valid():
+            form.save()
+            return redirect(list_course_url + '&msg=Successfully Added List!')
```
Comment (JackSekoranja) [inline cassdegrees/ui/views/lists.py]:
```suggestion
            return redirect(list_course_group_url + '&msg=Successfully Added List!')
```

File: cassdegrees/ui/views/lists.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+        if form.is_valid():
+            instance.lastUpdated = timezone.now().strftime('%Y-%m-%d')
+            instance.save(update_fields=['lastUpdated'])
+            form.save()
+            # POST Requests only carry boolean values over as string
+            # Only redirect the user to the list page if the user presses "Save and Exit".
+            # Otherwise, simply display a success message on the same page.
+            # todo: implement list view for admin page then change url
+            if request.POST.get('redirect') == 'true':
+                return redirect(list_program_url + '&msg=Successfully Edited List!')
```
Comment (JackSekoranja) [inline cassdegrees/ui/views/lists.py]:
```suggestion
                return redirect(list_course_group_url + '&msg=Successfully Edited List!')
```

**Review by JackSekoranja — CHANGES_REQUESTED** [review]
Okay, it works for me now. I do think it would be better to redirect to the course group listing page rather than programs or courses one (as creating a list would go to courses while editing one goes to programs). I've suggested changes below that will accomplish this, although they will all need to be accepted to work

**Comment by joshcbee** [comment]
Changes accepted and now redirecting to group listing page.

**Review by JackSekoranja — APPROVED** [review]
Looks good, happy for you to merge

**Pull request merged**